src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,526 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.AddImports;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.RemoveUnnecessaryImports;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.SyncNamespace
+{
+    internal abstract partial class AbstractSyncNamespaceService<TNamespaceDeclarationSyntax, TCompilationUnitSyntax>
+        where TNamespaceDeclarationSyntax : SyntaxNode
+        where TCompilationUnitSyntax : SyntaxNode 
+    {
+        /// <summary>
+        /// This code action tries to change the name of the namespace declaration to 
+        /// match the folder hierarchy of the document. The new namespace is constructed 
+        /// by concatenating the default namespace of the project and all the folders in 
+        /// the file path up to the project root.
+        /// 
+        /// For example, if he default namespace is `A.B.C`, file path is 
+        /// "[project root dir]\D\E\F\Class1.cs" and declared namespace in the file is
+        /// `Foo.Bar.Baz`, then this action will change the namespace declaration
+        /// to `A.B.C.D.E.F`. 
+        /// 
+        /// Note that it also handles the case where the target namespace or declared namespace 
+        /// is global namespace, i.e. default namespace is "" and the file is located at project 
+        /// root directory, and no namespace declaration in the document, respectively.
+        /// </summary>
+        internal sealed class ChangeNamespaceCodeAction : CodeAction
```

Comment:
this feels like this coudl be a SolutionChangeCodeAction.  But i'm not going to push hard on that.  But, basically, you don't seem to be utilizing any specific functionality of CodeAction to need to be a direct subclass.  i.e. you're not overriding specific preview behavior or anything like that.  All you're doing is returning a Solution instance.  So that can be represented more cleanly (IMO) with a SolutionChangeAction.

Reply:
Actually, one good reason to be a subclass is so that you can provide progress tracking using IProgressTracker.  We really should have that here.  This fixup can take a *long* time, and it would be good if our fixup steps updated the progress tracker with what's going on.